src/cmd/tools/dtest/docker/harness/query_api_test.go

```diff
@@ -0,0 +1,131 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package harness
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/m3db/m3/src/cmd/tools/dtest/docker/harness/resources"
+)
+
+var instantQueryBadRequestTest = []struct {
+	name string
+	q    string
+}{
+	// {"missing query", queryURL("query", "")},
+	// FAILING issue #2: invalid query string should result in 400
+	// {"invalid query", queryURL("query", "@!")},
+	{"invalid time", queryURL("time", "INVALID")},
+	{"invalid timeout", queryURL("timeout", "INVALID")},
+}
+
+func TestInvalidInstantQueryReturns400(t *testing.T) {
+	coord := singleDBNodeDockerResources.Coordinator()
+
+	for _, tt := range instantQueryBadRequestTest {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.NoError(t, coord.RunQuery(verifyStatus(400), tt.q), "for query '%v'", tt.q)
+		})
+	}
+}
+
+var queryBadRequestTest = []struct {
+	name string
+	q    string
+}{
+	{"missing query", queryRangeURL("query", "")},
+	// FAILING issue #2: invalid query string should result in 400
+	// {"invalid query", queryRangeURL("query", "@!")},
+	{"missing start", queryRangeURL("start", "")},
+	{"invalid start", queryRangeURL("start", "INVALID")},
+	{"missing end", queryRangeURL("end", "")},
+	{"invalid end", queryRangeURL("end", "INVALID")},
+	{"missing step", queryRangeURL("step", "")},
+	{"invalid step", queryRangeURL("step", "INVALID")},
+}
+
+func TestInvalidRangeQueryReturns400(t *testing.T) {
+	coord := singleDBNodeDockerResources.Coordinator()
+
+	for _, tt := range queryBadRequestTest {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.NoError(t, coord.RunQuery(verifyStatus(400), tt.q), "for query '%v'", tt.q)
+		})
+	}
+}
+
+func queryURL(key, value string) string {
+	params := map[string]string{
+		"query": "foo",
+	}
+
+	if value == "" {
+		delete(params, key)
+	} else {
+		params[key] = value
+	}
+
+	return "api/v1/query?" + queryString(params)
+}
+
+func queryRangeURL(key, value string) string {
+	params := map[string]string{
+		"query": "foo",
+		"start": "now",
+		"end":   "now",
+		"step":  "1000",
+	}
+
+	if value == "" {
+		delete(params, key)
+	} else {
+		params[key] = value
+	}
+
+	return "api/v1/query_range?" + queryString(params)
+}
+
+func queryString(params map[string]string) string {
+	p := make([]string, 0, len(params))
```

Comment:
Nit: let's not do these microoptimzations in test code and make it more readable instead:
```suggestion
	var p []string
```

Reply:
Simplified it a bit, though `var p []string` triggers a linting error:
```
src/cmd/tools/dtest/docker/harness/query_api_test.go:110:2: Consider preallocating `p` (prealloc)
        var p []string = nil
        ^
```

https://github.com/m3db/m3/pull/2873/commits/77f076c2efb862b562c7c349e6247d21a2ba78d3#diff-391a291f1f91098bc731847f8843207e5327fe87a52a4599a45ac0cf4c5732e7R111